File: tests/bindings/ruby/_coverage_to_lcov.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'set'

if ARGV.length < 2 || ARGV.length > 3
  warn "usage: #{$PROGRAM_NAME} <coverage-dir> <output-lcov> [source-root]"
  exit 2
end

coverage_dir = ARGV[0]
output_lcov = ARGV[1]
source_root = ARGV[2] ? File.expand_path(ARGV[2]) : nil

lib_root = source_root ? File.expand_path('ruby/lib', source_root) : nil

records = Hash.new do |hash, key|
  hash[key] = { found: Set.new, hits: Hash.new(0) }
end

normalize_path = lambda do |path|
  normalized = path.to_s
  return normalized unless source_root

  # Map bundled gem paths back to source-tree ruby/lib for stable reports.
  match = normalized.match(%r{/libsixel-ruby-[^/]+/lib/(.+)\z})
  if match
    candidate = File.expand_path(File.join('ruby/lib', match[1]), source_root)
    normalized = candidate if File.file?(candidate)
  end
  normalized
end

Dir.glob(File.join(coverage_dir, '*.marshal')).sort.each do |marshal_path|
  data = Marshal.load(File.binread(marshal_path))
  next unless data.is_a?(Hash)

  data.each do |path, line_hits|
    next unless line_hits.is_a?(Array)

    normalized = normalize_path.call(path)
    if lib_root
      next unless normalized.start_with?("#{lib_root}/")
    end

    entry = records[normalized]
    line_hits.each_with_index do |hit, idx|
      next if hit.nil?
      line = idx + 1
      entry[:found].add(line)
      entry[:hits][line] += hit if hit.positive?
    end
  end
end

File.open(output_lcov, 'w') do |f|
  records.keys.sort.each do |file_path|
    entry = records[file_path]
    next if entry[:found].empty?

    lines_found = 0
    lines_hit = 0

    f.puts 'TN:'
    f.puts "SF:#{file_path}"
    entry[:found].to_a.sort.each do |line|
      hit = entry[:hits][line]
      f.puts "DA:#{line},#{hit}"
      lines_found += 1
      lines_hit += 1 if hit.positive?
    end
    f.puts "LF:#{lines_found}"
    f.puts "LH:#{lines_hit}"
    f.puts 'end_of_record'
  end
end

File: tests/bindings/php/_coverage_to_lcov.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

if ($argc < 3 || $argc > 4) {
    fwrite(STDERR, "usage: {$argv[0]} <coverage-dir> <output-lcov> [source-root]\n");
    exit(2);
}

$coverageDir = $argv[1];
$outputLcov = $argv[2];
$sourceRoot = $argc === 4 ? realpath($argv[3]) : false;
$sourcePhpDir = is_string($sourceRoot) ? $sourceRoot . '/php/src/' : null;

$records = [];
$coverageFiles = glob($coverageDir . '/*.json');
if (!is_array($coverageFiles)) {
    $coverageFiles = [];
}
sort($coverageFiles);

foreach ($coverageFiles as $coverageFile) {
    $payload = file_get_contents($coverageFile);
    if (!is_string($payload) || $payload === '') {
        continue;
    }

    $decoded = json_decode($payload, true);
    if (!is_array($decoded)) {
        continue;
    }

    foreach ($decoded as $path => $lineHits) {
        if (!is_string($path) || !is_array($lineHits)) {
            continue;
        }

        $normalized = $path;
        if (is_string($sourceRoot)) {
            $candidate = null;
            if (is_string($sourcePhpDir) && strpos($path, $sourcePhpDir) === 0 && is_file($path)) {
                $candidate = $path;
            } elseif (preg_match('{/src/(.+\\.php)$}', $path, $matches) === 1) {
                $mapped = $sourceRoot . '/php/src/' . $matches[1];
                if (is_file($mapped)) {
                    $candidate = $mapped;
                }
            }

            if ($candidate === null) {
                continue;
            }
            $normalized = $candidate;

            if (is_string($sourcePhpDir) && strpos($normalized, $sourcePhpDir) !== 0) {
                continue;
            }
        }

        if (!isset($records[$normalized])) {
            $records[$normalized] = [
                'found' => [],
                'hits' => [],
            ];
        }

        foreach ($lineHits as $line => $hitCount) {
            if (!is_numeric($line) || !is_numeric($hitCount)) {
                continue;
            }
            $lineNumber = (int)$line;
            if ($lineNumber <= 0) {
                continue;
            }

            $hits = (int)$hitCount;
            if ($hits < 0) {
                $hits = 0;
            }

            $records[$normalized]['found'][$lineNumber] = true;
            if ($hits > 0) {
                if (!isset($records[$normalized]['hits'][$lineNumber])) {
                    $records[$normalized]['hits'][$lineNumber] = 0;
                }
                $records[$normalized]['hits'][$lineNumber] += $hits;
            }
        }
    }
}

ksort($records);

$out = fopen($outputLcov, 'wb');
if ($out === false) {
    fwrite(STDERR, "failed to open {$outputLcov} for writing\n");
    exit(1);
}

foreach ($records as $file => $entry) {
    $lines = array_keys($entry['found']);
    sort($lines, SORT_NUMERIC);
    if ($lines === []) {
        continue;
    }

    $linesFound = 0;
    $linesHit = 0;

    fwrite($out, "TN:\n");
    fwrite($out, "SF:{$file}\n");

    foreach ($lines as $lineNumber) {
        $hits = $entry['hits'][$lineNumber] ?? 0;
        fwrite($out, "DA:{$lineNumber},{$hits}\n");
        $linesFound++;
        if ($hits > 0) {
            $linesHit++;
        }
    }

    fwrite($out, "LF:{$linesFound}\n");
    fwrite($out, "LH:{$linesHit}\n");
    fwrite($out, "end_of_record\n");
}

fclose($out);
